File: semgus-sketch/src/main/java/org/semgus/sketch/util/Evaluator.java
package org.semgus.sketch.util;

import org.semgus.java.object.SmtTerm;
import org.semgus.sketch.base.Expr;
import org.semgus.sketch.base.Id;
import org.semgus.sketch.base.Param;

import java.util.List;
import java.util.Map;
import java.util.function.Function;

public interface Evaluator {
  Map<String, Function<List<Expr>, Expr>> fns = Map.ofEntries(
    // Core Theory
    Map.entry("true", es -> Expr.refPlain("true")),
    Map.entry("false", es -> Expr.refPlain("false")),
    // TODO: Support Expr.Nary.
    Map.entry("and", es -> Expr.binaryPlain("&&", es.get(0), es.get(1))),
    Map.entry("or", es -> Expr.binaryPlain("||", es.get(0), es.get(1))),
    Map.entry("not", es -> Expr.unaryPlain("!", es.get(0))),
    Map.entry("!", es -> Expr.unaryPlain("!", es.get(0))),
    Map.entry("xor", es -> Expr.binaryPlain("^", es.get(0), es.get(1))),
    // TODO: Map.entry("=>")?
    Map.entry("=", es -> Expr.binaryPlain("==", es.get(0), es.get(1))),
    // TODO: Map.entry("distinct")?
    Map.entry("ite", es -> Expr.cond(es.get(0), es.get(1), es.get(2))),

    // Ints Theory
    Map.entry("-", es -> es.size() == 1
      ? Expr.unaryPlain("-", es.get(0))
      : Expr.binaryPlain("-", es.get(0), es.get(1)
    )),
    Map.entry("+", es -> Expr.binaryPlain("+", es.get(0), es.get(1))),
    Map.entry("*", es -> Expr.binaryPlain("*", es.get(0), es.get(1))),
    Map.entry("div", es -> Expr.binaryPlain("/", es.get(0), es.get(1))),
    Map.entry("mod", es -> Expr.binaryPlain("%", es.get(0), es.get(1))),
    // TODO: Map.entry("abs")? Use Expr.Cond.
    Map.entry("<=", es -> Expr.binaryPlain("<=", es.get(0), es.get(1))),
    Map.entry("<", es -> Expr.binaryPlain("<", es.get(0), es.get(1))),
    Map.entry(">=", es -> Expr.binaryPlain(">=", es.get(0), es.get(1))),
    Map.entry(">", es -> Expr.binaryPlain(">", es.get(0), es.get(1)))

    // TODO: Bit Vectors Theory

    // TODO: Strings Theory
  );

  static Expr eval(SmtTerm t) {
    return switch (t) {
      case SmtTerm.Application tApp -> fns.getOrDefault(
          tApp.name().name(),
          es -> Expr.appPlain(tApp.name().name(), es))
        .apply(tApp.arguments().stream()
          .map(SmtTerm.Application.TypedTerm::term)
          .map(Evaluator::eval)
          .toList());
      case SmtTerm.Variable tVar -> Expr.refPlain(tVar.name());
      case SmtTerm.CNumber tNum -> Expr.refPlain(tNum.value());
      case SmtTerm.CString tStr -> Expr.refPlain(tStr.value());
      // TODO: What about exists?
      case SmtTerm.Quantifier tQnt -> tQnt.type().toString().equals("∀")
        ? Expr.forall(
        tQnt.bindings().stream()
          .map(typedVar -> Param.plain(
            Id.fromSemgusId(typedVar.type()),
            typedVar.name()))
          .toList(),
        eval(tQnt.child()))
        : Expr.refPlain("??");
      // TODO: case SmtTerm.CBitVector ->
      default -> throw new IllegalStateException("Unexpected value: " + t);
    };
  }
}

File: semgus-sketch/src/main/java/org/semgus/sketch/util/Evaluator.kt
package org.semgus.sketch.util

import org.semgus.java.`object`.SmtTerm
import org.semgus.java.`object`.SmtTerm.Application
import org.semgus.java.`object`.SmtTerm.CNumber
import org.semgus.java.`object`.SmtTerm.CString
import org.semgus.java.`object`.SmtTerm.Quantifier
import org.semgus.java.`object`.SmtTerm.Variable
import org.semgus.sketch.syntax.*
import org.semgus.sketch.syntax.Op.*

/**
 * Evaluates SMT terms to Sketch expressions.
 */
internal fun SmtTerm.toExpr(): Expr = when (this) {
  is Application -> (fns[this.name().name()] ?: TODO("Not supported theory $this"))
    .invoke(this.arguments().asSequence().map { it.term().toExpr() })
  is Variable -> refPlain(this.name())
  is CNumber -> refPlain(this.value())
  is CString -> refPlain(this.value())
  // TODO: What about exists?
  is Quantifier -> forall(
    binds = this.bindings().asSequence().map { typedVar -> param(typedVar) },
    e = this.child().toExpr(),
  )
  else -> throw IllegalStateException("Unexpected value: $this")
}

/**
 * The helper map for evaluation.
 */
internal val fns = mutableMapOf<String, (Sequence<Expr>) -> Expr>(
  // Core Theory
  "true" to { refPlain("true") },
  "false" to { refPlain("false") },
  "and" to { es -> nary(AND, es) },
  "or" to { es -> nary(OR, es) },
  "not" to { (iter) -> unary(NOT, iter.next()) },
  "!" to { (iter) -> unary(NOT, iter.next()) },
  "xor" to { (iter) -> binary(XOR, iter.next(), iter.next()) },
  // TODO: "=>"?
  "=" to { (iter) -> binary(EQ, iter.next(), iter.next()) },
  // TODO: "distinct"?
  "ite" to { (iter) -> ite(iter.next(), iter.next(), iter.next()) },

  // Ints Theory
  "-" to { (iter) ->
    val next = iter.next()
    if (!iter.hasNext()) unary(MINUS, next)
    else binary(MINUS, next, iter.next())
  },
  "+" to { (iter) -> binary(PLUS, iter.next(), iter.next()) },
  "*" to { (iter) -> binary(TIMES, iter.next(), iter.next()) },
  "div" to { (iter) -> binary(DIV, iter.next(), iter.next()) },
  "mod" to { (iter) -> binary(MOD, iter.next(), iter.next()) },
  "abs" to { (iter) ->
    val next = iter.next()
    ite(binary(LT, next, refPlain(0)), unary(MINUS, next), next)
  },
  "<=" to { (iter) -> binary(LTE, iter.next(), iter.next()) },
  "<" to { (iter) -> binary(LT, iter.next(), iter.next()) },
  ">=" to { (iter) -> binary(GTE, iter.next(), iter.next()) },
  ">" to { (iter) -> binary(GT, iter.next(), iter.next()) },

  // TODO: Bit Vectors Theory

  // TODO: Strings Theory
)

/**
 * Makes destructuring a sequence in lambda possible.
 */
private operator fun <T> Sequence<T>.component1(): Iterator<T> {
  return this.iterator()
}
